File: src/1768-merge-strings-alternately.ts
import { check } from "./utils/check"

function mergeAlternately(word1: string, word2: string): string {
    const A = word1.length
    const B = word2.length
    let a = 0
    let b = 0
    const result: string[] = []

    let word = 1
    while (a < A && b < B) {
        if (word === 1) {
            result.push(word1[a])
            a++
            word = 2
        } else {
            result.push(word2[b])
            b++
            word = 1
        }
    }
    while (a < A) {
        result.push(word1[a])
        a++
    }
    while (b < B) {
        result.push(word2[b])
        b++
    }
    return result.join("")
};


// ---- tests ----
{

  check("case1 same length", mergeAlternately("abc", "pqr"), "apbqcr")
  check("case2 word2 longer", mergeAlternately("ab", "pqrs"), "apbqrs")
  check("case3 word1 longer", mergeAlternately("abcd", "pq"), "apbqcd")
  check("case4 [abc, defgh]", mergeAlternately("abc", "defgh"), "adbecfgh")
}

// make this file a module so its declarations stay file-scoped
export {}

File: src_py/1768-merge-strings-alternately.py
"""
1768. Merge Strings Alternately
Difficulty: Easy

Merge two strings by adding letters in alternating order, starting with word1.
If a string is longer than the other, append the additional letters onto the
end of the merged string.
"""


def merge_alternately(word1: str, word2: str) -> str:
    result = []
    i = j = 0
    turn = 0
    while i < len(word1) and j < len(word2):
        if turn == 0:
            result.append(word1[i])
            i += 1
            turn = 1
        else:
            result.append(word2[j])
            j += 1
            turn = 0
    result.append(word1[i:])
    result.append(word2[j:])
    return "".join(result)


if __name__ == "__main__":
    assert merge_alternately("abc", "pqr") == "apbqcr"
    assert merge_alternately("ab", "pqrs") == "apbqrs"
    assert merge_alternately("abcd", "pq") == "apbqcd"
    assert merge_alternately("abc", "defgh") == "adbecfgh"
    print("1768 OK:", merge_alternately("abc", "defgh"))
